### PR TITLE
drivers/sht2x: some small improvements

### PR DESCRIPTION
**Pull request by gschorcht**
### Contribution description

This PR provides the following improvements for the SHT2x driver:

- migration to `ztimer` 8a605517f5343532a4cc25866c9bd18818f17d50 and 367549940de3bd8910052334c34af028d4992741
- floating point arithmetics replaced by integer arithmetics f424caebbfec9f2be56aa2337c6cc09dba5b97d6
- fix of sleep times (typical times replaced by maximum times as recommended by the datasheet) 13615c72094b8541ee62c3e8ed5a36dc4d725fd0
- release of the I2C bus during sleep 9415c216cfab734520ef98dd00b350c22f342c60
- Kconfig configuration of sensor parameters added dadbbcb4c328350893db53ba6743d03cb34ecc1c
- no-hold mode is now the default mode instead of the hold mode

Regarding the sleep times: The typical measurement times were used as sleep times. According to the datasheet, typical measurements are only recommended for calculating energy consumption, while maximum values should be used for calculating waiting times in communication. Therefore, the typical time values were replaced by maximum time values for the sleep in no-hold mode.

Regarding the hold mode: In hold mode, the sensor uses clock stretching until the measurement results can be read by the MCU. This blocks both the I2C bus and the MCU during the entire measurement, which can take up to 85 ms if I2C is not interrupt-driven. Therefore, the no-hold mode is now used by default, where the calling thread sleeps during the measurement, but the MCU is not completely blocked. Furthermore, the hold mode requires that the MCU supports clock stretching. Even if the MCU supports clock stretching, the hold mode with clock stretching doesn't seem to work with different MCUs. I couldn't get it working for STM32 and ESP32.

Regarding the I2C bus during sleep: If the no-hold mode is used and the calling thread sleeps up to 85 ms, it makes sense to release the I2C bus until the measurement results are available.

### Testing procedure

`tests/driver_sht2x` should still work.

### CODE REVIEW

**Review by benpicco — COMMENTED** [review]
Looks good to me.
While you're at it, do you also want to replace `sht2x_checkcrc()` with `crc8()` from `checksum/crc8`?
It's pretty much the same code.

File: drivers/sht2x/sht2x.c [hunk 1]
```diff
@@ -407,29 +408,13 @@ static int read_sensor_poll(const sht2x_t* dev, cmd_t command, uint16_t *val)
     return SHT2X_OK;
 }
 
-static const uint16_t POLYNOMIAL = 0x131;       /* P(x)=x^8+x^5+x^4+1 = 100110001 */
+static const uint8_t POLYNOMIAL = 0x31;       /* P(x)=x^8+x^5+x^4+1 = 100110001 */
```
Comment (benpicco) [inline drivers/sht2x/sht2x.c:411]: Why is this `0x31` now?

**Review by benpicco — APPROVED** [review]
Please squash

**Comment by gschorcht** [comment]
BTW, while investigating the problem why the HOLD measurement mode does not work on stm32f429zi and stm32f767zi, I found that the 16-bit timeout counter `TICK_TIMEOUT` https://github.com/RIOT-OS/RIOT/blob/10c1909f2dfd8d751db26087bd9b22bf689dcaa5/cpu/stm32/periph/i2c_2.c#L55 is too small and only allows clock stretching up to 15 ms. However, the SHT2x sensor stretches the clock in HOLD mode for up to 85 ms during the measurement.

The question is whether the STM32 I2C implementation should better use a 32-bit count and set the start value of the timout counter to `0x0008ffff` to allow a clock stretch of about 100 ms.

On the other hand, using HOLD mode and clock stretching for up to 85 ms is a rather bad idea, as this not only blocks the I2C bus, but also the MCU if the I2C implementation is not interrupt driven. So I wonder if we should remove the HOLD mode support completely.

It is also interesting that the `si70xx` driver, which is for an almost identical sensor, which should actually also work with the `sht2x` driver, only supports HOLD mode and nothing else :worried:

**Comment by gschorcht** [comment]
> BTW, while investigating the problem why the HOLD measurement mode does not work on stm32f429zi and stm32f767zi, I found that the 16-bit timeout counter `TICK_TIMEOUT`

We have the same problem with ESPs. In the I2C specification there is no limit for the clock stretching, while the SMBus specification limits it to 35 ms.

**Comment by gschorcht** [comment]
@benpicco Squashed :smile:

**Comment by benpicco** [comment]
It's a bit wild that the I2C driver uses a busy loop to count down the the timeout, but then again it also does a busy polling on the I2C registers instead of waiting for an interrupt.

What's the advantage to HOLD mode anyway? Is it to simplify or speed up the communication?

bors merge

**Comment by gschorcht** [comment]
Thanks for reviewing and merging.

**Pull request merged**